pkg/operator/encryption/controllers/migrators/inprocess_processor.go

```diff
@@ -0,0 +1,178 @@
+		for {
+			allResource, err := p.dynamicClient.Resource(gvr).List(opts)
+			if err != nil {
+				klog.Warningf("List of %v failed: %v", gvr, err)
+				if errors.IsResourceExpired(err) {
+					token, err := inconsistentContinueToken(err)
+					if err != nil {
+						return nil, err
+					}
+					opts.Continue = token
+					klog.V(4).Infof("Relisting %v after handling expired token", gvr)
+					continue
+				} else if retryable := canRetry(err); retryable == nil || *retryable == false {
+					return nil, err // not retryable or we don't know. Return error and controller will restart migration.
+				} else {
+					if seconds, delay := errors.SuggestsClientDelay(err); delay {
+						time.Sleep(time.Duration(seconds) * time.Second)
+					}
+					klog.V(4).Infof("Relisting %v after retryable error: %v", gvr, err)
+					continue
+				}
+			}
+
+			migrationStarted := time.Now()
+			klog.V(4).Infof("Migrating %d objects of %v", len(allResource.Items), gvr)
+			err = p.processList(allResource)
+			if err != nil {
+				errs = append(errs, err)
+			}
+			klog.V(4).Infof("Migration of %d objects of %v finished in %v", len(allResource.Items), gvr, time.Now().Sub(migrationStarted))
+
+			allResource.Items = nil // do not accumulate items, this fakes the visitor pattern
+			return allResource, nil // leave the rest of the list intact to preserve continue token
+		}
+	}))
+
+	migrationStarted := time.Now()
+	listPager.FullListIfExpired = false // prevent memory explosion from full list
+	_, listErr := listPager.List(p.ctx, metav1.ListOptions{})
+	klog.V(4).Infof("Migration for %v finished in %v", gvr, time.Now().Sub(migrationStarted))
+	errs = append(errs, listErr)
+	return utilerrors.NewAggregate(errs)
+}
+
+func (p *ListProcessor) processList(l *unstructured.UnstructuredList) error {
+	workCh := make(chan *unstructured.Unstructured, p.concurrency)
+	go func() {
+		defer utilruntime.HandleCrash()
+		defer close(workCh)
+		for i := range l.Items {
+			select {
+			case workCh <- &l.Items[i]:
+			case <-p.ctx.Done():
+				return
+			}
+		}
+	}()
+
+	var wg sync.WaitGroup
+	wg.Add(p.concurrency)
+	errCh := make(chan error)
+	for i := 0; i < p.concurrency; i++ {
+		go func() {
+			defer utilruntime.HandleCrash()
+			defer wg.Done()
+			p.worker(workCh, errCh)
+		}()
+	}
+
+	go func() {
+		defer utilruntime.HandleCrash()
+		wg.Wait()
+		close(errCh)
+	}()
+
+	var errors []error
+	for err := range errCh {
+		errors = append(errors, err)
+	}
+	return utilerrors.NewAggregate(errors)
+}
+
+func (p *ListProcessor) worker(workCh <-chan *unstructured.Unstructured, errCh chan<- error) {
+	for item := range workCh {
+		err := executeWorkerFunc(p.workerFn, item)
+		if err != nil {
+			select {
+			case errCh <- err:
+				continue
+			case <-p.ctx.Done():
+				return
+			}
+		}
+	}
+}
+
+func executeWorkerFunc(workerFn WorkerFunc, obj *unstructured.Unstructured) (result error) {
+	defer func() {
+		if r := recover(); r != nil {
```

Comment:
isn't there something like util.HandlePanic() ?

Reply:
not if you want to capture the error and return it, notice that this function makes use of `named return value`, panic will be captured, transformed to an error and returned to the caller.